apm-agent-plugins/apm-asynchttpclient-plugin/src/main/java/co/elastic/apm/agent/asynchttpclient/AsyncHttpClientInstrumentation.java

```diff
@@ -0,0 +1,153 @@
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        public static void onBeforeExecute(@Advice.Argument(value = 0, readOnly = true) Request request,
+                                           @Advice.Argument(value = 1, readOnly = false) AsyncHandler asyncHandler,
+                                           @Advice.Local("span") Span span) {
+            if(tracer == null || tracer.getActive() == null) {
+                return;
+            }
+
+            final TraceContextHolder<?> parent = tracer.getActive();
+            span = HttpClientHelper.startHttpClientSpan(parent, request.getMethod(), request.getUri().toFullUrl(), request.getUri().getHost());
+
+            if(span != null) {
+                span.activate();
+                request.getHeaders().add(TraceContext.TRACE_PARENT_HEADER, span.getTraceContext().getOutgoingTraceParentHeader().toString());
+            }
+            if(!request.getHeaders().contains(TraceContext.TRACE_PARENT_HEADER) && parent != null) {
+                request.getHeaders().add(TraceContext.TRACE_PARENT_HEADER, parent.getTraceContext().getOutgoingTraceParentHeader().toString());
+            }
+
+            // TODO: Optimize the new call away
+            asyncHandler = new AsyncHandlerWrapperCreator().wrap(asyncHandler, span, request);
+        }
+
+        @Nullable
+        @VisibleForAdvice
+        @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class)
+        public static void onAfterExecute(@Advice.Local("span") Span span,
+                                          @Advice.Thrown Throwable t) {
+            if(span != null) {
+                span.deactivate();
+            }
+        }
+    }
+
+    public static class AsyncHandlerWrapperCreator {
+        public AsyncHandler<Object> wrap(final AsyncHandler asyncHandler, Span span, Request originalRequest) {
+            return new AsyncHandlerWrapper(asyncHandler, span, originalRequest);
+        }
+
+        private static class AsyncHandlerWrapper implements AsyncHandler {
+            private Span span;
+            private AsyncHandler delegate;
+            private Request originalRequest;
+
+            public AsyncHandlerWrapper(AsyncHandler delegate, Span span, Request originalRequest) {
+                this.span = span;
+                this.delegate = delegate;
+                this.originalRequest = originalRequest;
+            }
+
+            @Override
+            public State onStatusReceived(HttpResponseStatus responseStatus) throws Exception {
+                span.getContext().getHttp().withStatusCode(responseStatus.getStatusCode());
+                return delegate.onStatusReceived(responseStatus);
+            }
+
+            @Override
+            public State onHeadersReceived(HttpHeaders headers) throws Exception {
+                if(!headers.contains(TraceContext.TRACE_PARENT_HEADER)) {
+                    headers.add(TraceContext.TRACE_PARENT_HEADER, "");
```

Comment:
I think this is a callback when response headers are received and thus can be removed.